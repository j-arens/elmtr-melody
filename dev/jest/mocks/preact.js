module.exports = {
    unmountComponent: jest.fn(),
};
