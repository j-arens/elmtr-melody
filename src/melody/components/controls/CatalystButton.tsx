import BaseButton from '@components/BaseButton/';
import Icon from '@components/Icon/';
import { WithOptionalClassName } from '@melody/components/type';
import { cycleState } from '@redux/actions';
import { Action, State } from '@redux/type';
import stateMachine from '@state-machine/index';
import { MachineAction, MachineStates } from '@state-machine/type';
import { formatPlaybackRate } from '@utils/index';
import { h } from 'preact';
import { connect } from 'preact-redux';
const s = require('./styles.scss');

const mapStateToProps = (state: State) => ({
    currentState: state.currentState,
    playbackRate: state.playbackRate,
});

const mapDispatchToProps = dispatch => ({
    cycleState: (action: MachineAction) => dispatch(cycleState(action)),
});

interface Props extends WithOptionalClassName {
    currentState: MachineStates;
    playbackRate: number;
    cycleState: (action: MachineAction) => Action;
}

function getMachineAction(currentState: MachineStates) {
    switch (currentState) {
        case 'buffering': {
            return 'NOOP';
        }
        case 'playing': {
            return 'STOP';
        }
        case 'stopped': {
            return 'PLAY';
        }
        default: {
            return 'FAILED';
        }
    }
}

const CatalystButton = ({
    playbackRate,
    currentState,
    cycleState,
    className = '',
}: Props) => {
    const defaultClass = `${className} ${s.playbackCtrl} melody-playbackCtrl`;
    const bufferClass = `${defaultClass} ${s['playbackCtrl--buffering']}`;
    const cycle = () => cycleState(getMachineAction(currentState));
    return (currentState === 'buffering' ?
        <BaseButton
            key="catalyst"
            className={`${bufferClass} melody-controlBar__catalyst`}
            data-playbackRate=""
        >
            <Icon className={s.playbackCtrl__icon} name="buffer" />
        </BaseButton>
        :
        <BaseButton
            key="catalyst"
            className={`${defaultClass} melody-controlBar__catalyst`}
            onClick={cycle}
            data-playbackRate={formatPlaybackRate(playbackRate)}
        >
            <Icon
                className={s.playbackCtrl__icon}
                name={currentState !== 'playing' ? 'play' : 'pause'}
            />
        </BaseButton>
    );
};

export default connect(mapStateToProps, mapDispatchToProps)(CatalystButton);
