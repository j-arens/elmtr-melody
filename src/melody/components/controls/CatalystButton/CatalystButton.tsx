import BaseButton from '@components/BaseButton/';
import Icon from '@components/Icon/';
import { WithOptionalClassName } from '@melody/components/type';
import { Action } from '@redux/type';
import stateMachine from '@state-machine/index';
import { MachineAction, MachineStates } from '@state-machine/type';
import {
    formatPlaybackRate,
    getNextMachineAction,
} from '@utils/index';
import { h } from 'preact';
const s = require('../styles.scss');

export interface StateProps {
    currentState: MachineStates;
    playbackRate: number;
}

export interface DispatchProps {
    cycleState: (action: MachineAction) => Action;
}

type Props = StateProps & DispatchProps & WithOptionalClassName;

export default ({
    playbackRate,
    currentState,
    cycleState,
    className = '',
}: Props) => {
    const defaultClass = `${className} ${s.playbackCtrl} melody-playbackCtrl`;
    const bufferClass = `${defaultClass} ${s['playbackCtrl--buffering']}`;
    const cycle = () => cycleState(getNextMachineAction(currentState));
    return (currentState === 'buffering' ?
        <BaseButton
            key="catalyst"
            className={`${bufferClass} melody-controlBar__catalyst`}
            data-playbackRate=""
        >
            <Icon className={s.playbackCtrl__icon} name="buffer" />
        </BaseButton>
        :
        <BaseButton
            key="catalyst"
            className={`${defaultClass} melody-controlBar__catalyst`}
            onClick={cycle}
            data-playbackRate={formatPlaybackRate(playbackRate)}
        >
            <Icon
                className={s.playbackCtrl__icon}
                name={currentState !== 'playing' ? 'play' : 'pause'}
            />
        </BaseButton>
    );
};
