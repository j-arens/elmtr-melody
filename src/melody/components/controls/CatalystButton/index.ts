import { WithOptionalClassName } from '@components/type';
import { cycleState } from '@redux/modules/machine/actions';
import { State } from '@redux/type';
import { MachineAction } from '@state-machine/type';
import { connect } from 'preact-redux';
import CatalystButton, { DispatchProps, StateProps } from './CatalystButton';

const mapState = (state: State) => ({
    currentState: state.machine.currentState,
    playbackRate: state.audio.playbackRate,
});

const mapDispatch = dispatch => ({
    cycleState: (action: MachineAction) => dispatch(cycleState(action)),
});

export default connect<StateProps, DispatchProps, WithOptionalClassName>(mapState, mapDispatch)(CatalystButton);
