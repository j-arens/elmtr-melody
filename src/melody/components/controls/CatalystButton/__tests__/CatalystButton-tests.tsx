import { h } from 'preact';
import { render } from 'preact-render-to-string';
import CatalystButton from '../CatalystButton';

describe('CatalystButton', () => {
    it('renders with a play icon if current state is stopped', () => {
        const component = render(
            <CatalystButton
                currentState="stopped"
                className="carrera-white"
            />,
        );
        expect(component).toMatchSnapshot();
    });

    it('renders with a pause icon if current state is playing', () => {
        const component = render(
            <CatalystButton
                currentState="playing"
                className="carrera-white"
            />,
        );
        expect(component).toMatchSnapshot();
    });

    it('renders with a buffering icon if current state is buffering', () => {
        const component = render(
            <CatalystButton
                currentState="buffering"
                className="carrera-white"
            />,
        );
        expect(component).toMatchSnapshot();
    });
});
