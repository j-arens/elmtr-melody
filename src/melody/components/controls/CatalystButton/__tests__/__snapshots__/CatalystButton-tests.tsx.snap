// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalystButton renders with a buffering icon if current state is buffering 1`] = `
<button data-playbackRate class="btn carrera-white playbackCtrl melody-playbackCtrl playbackCtrl--buffering melody-controlBar__catalyst"
role="button">
  <svg class="playbackCtrl__icon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
    <g>
      <circle class="melody-playbackCtrl__ring" cx="12" cy="12" r="12"></circle>
      <g class="melody-playbackCtrl__buffer-dots">
        <circle cx="4" cy="13" r="2"></circle>
        <circle cx="12" cy="13" r="2"></circle>
        <circle cx="20" cy="13" r="2"></circle>
      </g>
    </g>
  </svg>
</button>
`;

exports[`CatalystButton renders with a pause icon if current state is playing 1`] = `
<button data-playbackRate="xundefined" class="btn carrera-white playbackCtrl melody-playbackCtrl melody-controlBar__catalyst"
role="button">
  <svg class="playbackCtrl__icon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
    <g>
      <circle class="melody-playbackCtrl__ring" cx="12" cy="12" r="12"></circle>
      <path d="M6 19h4V5H6v14zm8-14v14h4V5h-4z"></path>
    </g>
  </svg>
</button>
`;

exports[`CatalystButton renders with a play icon if current state is stopped 1`] = `
<button data-playbackRate="xundefined" class="btn carrera-white playbackCtrl melody-playbackCtrl melody-controlBar__catalyst"
role="button">
  <svg class="playbackCtrl__icon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
    <g>
      <circle class="melody-playbackCtrl__ring" cx="12" cy="12" r="12"></circle>
      <path d="M8 5v14l11-7z"></path>
    </g>
  </svg>
</button>
`;
